lxd/container_lxc.go

```diff
@@ -1440,32 +1441,42 @@ func (c *containerLXC) expandDevices() error {
 		for k, v := range profileDevices {
 			devices[k] = v
 
-			if v["type"] == "disk" && v["path"] == "/" {
-				rootDevices++
-				profileStoragePool = v["pool"]
+			if isRootDiskDevice(v) {
+				if !shared.StringInSlice(v["pool"], profileRootDiskDevices) {
+					profileRootDiskDevices = append(profileRootDiskDevices, v["pool"])
+				}
 			}
-
 		}
 	}
 
 	// Stick local devices on top
 	for k, v := range c.localDevices {
 		devices[k] = v
 
-		if v["type"] == "disk" && v["path"] == "/" {
+		if isRootDiskDevice(v) {
 			c.storagePool = v["pool"]
 		}
 	}
 
-	if rootDevices > 1 {
-		return fmt.Errorf("Failed to detect unique root device: Multiple root devices detected.")
+	// In case the container relies on profiles for its root device there
+	// can only be one.
+	if len(profileRootDiskDevices) > 1 && c.storagePool == "" {
+		if checkRootDisk {
+			return fmt.Errorf("Failed to detect unique root device: Multiple root devices detected.")
+		} else {
+			shared.LogWarnf("Failed to detect unique root device: Multiple root devices detected.")
+		}
 	}
 
 	if c.storagePool == "" {
-		if profileStoragePool == "" {
+		if profileRootDiskDevices != nil {
+			c.storagePool = profileRootDiskDevices[0]
+		}
+		if checkRootDisk && c.storagePool == "" {
 			return fmt.Errorf("No storage pool specified.")
+		} else if c.storagePool == "" {
+			shared.LogWarnf("No storage pool specified.")
 		}
-		c.storagePool = profileStoragePool
 	}
 
 	c.expandedDevices = devices
```

Comment:
So I really don't think we want expandConfig or expandDevices to do ANY config/devices validation.
Those functions are called constantly from all over the place and should be very minimal and very quick.

The only times where the container's configuration should be validate is when the container is created or when the container is changed. If we do it at any other time, we risk re-introducing those kind of problems where even trivial things like "lxc config show" now fail, making it impossible for the user to recover unless they do DB surgery.